tls/s2n_psk.c

```diff
@@ -327,8 +329,71 @@ int s2n_connection_set_external_psks(struct s2n_connection *conn, struct s2n_ext
         GUARD(s2n_psk_init(new_psk, S2N_PSK_TYPE_EXTERNAL));
         GUARD(s2n_psk_new_identity(new_psk, psk_vec[i].identity, psk_vec[i].identity_length));
         GUARD(s2n_psk_new_secret(new_psk, psk_vec[i].secret, psk_vec[i].secret_length));
-        GUARD(s2n_psk_set_hmac(new_psk, psk_vec[i].hmac));
+        GUARD_AS_POSIX(s2n_psk_set_hmac(new_psk, psk_vec[i].hmac));
     }
 
     return S2N_SUCCESS;
 }
+
+/* Match a PSK identity received from the client against the server's known PSK identities.
+ *
+ * While both the client's offered identities and whether a match was found are public, we should make an attempt
+ * to keep the server's known identities a secret. We will make comparisons to the server's identities constant
+ * time (to hide partial matches) and not end the search early when a match is found (to hide the ordering).
+ *
+ * Keeping these comparisons constant time is not high priority. There's no known attack using these timings,
+ * and an attacker could probably guess the server's known identities just by observing the public identities
+ * sent by clients.
+ */
+S2N_RESULT s2n_match_psk_identity(struct s2n_array *known_psks, const struct s2n_blob *wire_identity,
+        struct s2n_psk **match)
+{
+    ENSURE_REF(match);
+    ENSURE_REF(wire_identity);
+    ENSURE_REF(known_psks);
+
+    *match = NULL;
+
+    for (size_t i = 0; i < known_psks->len; i++) {
+        struct s2n_psk *psk = NULL;
+        GUARD_RESULT(s2n_array_get(known_psks, i, (void**)&psk));
+        ENSURE_REF(psk);
+
+        ENSURE_REF(psk->identity.data);
+        ENSURE_REF(wire_identity->data);
+
+        uint32_t compare_size = MIN(wire_identity->size, psk->identity.size);
+        if (s2n_constant_time_equals(psk->identity.data, wire_identity->data, compare_size)
+            & (psk->identity.size == wire_identity->size) & (!*match)) {
+            *match = psk;
+        }
+    }
+    return S2N_RESULT_OK;
+}
+
+int s2n_select_psk_identity(struct s2n_connection *conn, 
+                   struct s2n_psk_identity *identities, size_t identities_length,
+                   uint16_t *chosen_wire_index)
+{
```

Comment:
Yes, I think special casing the default case would be an efficient way to go about the search. The reason I choose to do the extra lookup was because strictly speaking the callback is meant to only return the `chosen_wire_index` and not set it on the connection `conn->psk_params.chosen_psk`.  Though in this case, we could simplify the lookups.

Reply:
Yes, I think special casing the default case would be an efficient way to go about the search, though it is worthy to point out that we would still have to do this lookup for any customer provided callback. The reason I choose to do the extra lookup was because strictly speaking the callback is meant to only return the `chosen_wire_index` and not set it on the connection `conn->psk_params.chosen_psk`.